src/GLSL.hs

```diff
@@ -0,0 +1,75 @@
+module GLSL where
+import Syntax (OpaqueType)
+
+data GSize = Two
+           | Three
+           | Four
+        deriving (Eq, Ord)
+
+data GType = Void
+           | GBool
+           | GInt
+           | GUInt
+           | GFloat
+           | GDouble
+           | GVec GSize GType
+           | GMat GSize GSize
+
+data GOp = GArr
+         | GField
+         | GMult
+         | GMod
+         | GDiv
+         | GAdd
+         | GSub
+         | GRShift
+         | GLShift
+         | GLt
+         | GLte
+         | GGt
+         | GGte
+         | GEq
+         | GNEq
+         | GAnd
+         | GOr
+         | GXOr
+         | GBAnd
+         | GBOr
+         | GBXOr
+
+data GUOp = GFLip
+          | GNot
+          | GNeg
+          | GPos
+
+data GConst = GR String
+            | GI Int
+            | GB Bool
+            | GF Float
+            | GD Double
+            | GV [GConst]
+            | GM [[GConst]]
+
+data GExpr = GCall String [GExpr]
+           | GParen GExpr
+           | GAssign String GExpr
+           | Imm GConst
+           | GUn GUOp GExpr
+           | GBin GOp GExpr GExpr
+           | GTern GExpr GExpr GExpr
+
+data GStmt = GFor GExpr GExpr GExpr [GStmt]
+           | GWhile GExpr [GStmt]
+           | GDoWhile GExpr [GStmt]
+           | GIf [(Maybe GExpr, [GStmt])] -- This is so that we can have if/else if/else chains
+           | GSwitch GExpr [(GConst,[GStmt])]
+           | GStmt GExpr
+           | GRet GExpr
+           | GBreak
+           | GContinue
+
+data GTop = GFunc String [(String,GType)] GType [GStmt]
+          | GGlob GExpr                     -- This is for global variables
```

Comment:
I believe in GLSL (i may have to check this) you can only declare globals as names with associated types with or without a corresponding binding. With the way this is written out we can write arbitrary expressions like binary ops at the top level w/out being part of a binding. What do you think if we created a new category for global vars kinda like this (also a rough idea).
```haskell
-- top level declarations, funcs, global vars, or opaque types.
data GTop = GGlob Global | ...

-- globals we can write out
data Global = GlobalDecl Name GType -- declarations w/ no binding
                     | GlobalDef Name GType GExpr  -- declarations w/ binding (a def)
```

Reply:
I think this begs the question as to how much of the full specification do we want to encompass?
I agree with you and was wondering that myself when I was writing it out, but my thought was we could enforce that it's only assignments at the type level (but that's probably poor form, I know).